### PR TITLE
Queue 'new post' notification

### LINKED ISSUE #1869: [Subscriptions] Queue "new post" email
**Relates to #978.**

A nice first use-case for playing around with our new queueing support (#1773).

1. Send `new_post` email via queue (which uses the `sync` driver by default, so this should not change behavior).
2. Document the setup for real queue workers using e.g. Redis, and try it out on discuss.

Once this is implemented, we can start showcasing this to extension developers, and tackle other tasks in the system that could benefit from being offloaded to the queue.

### PR DESCRIPTION
**Pull request by dsevillamartin**
**Fixes flarum/core#1869**

The last post number needs to be passed because if using a queue driver other than `sync`, the discussion's details are updated (or another post could have been added) so it will target the current or later posts. This results in no user ever having read the latest post in the discussion and not qualifying for a notification. Probably wouldn't be needed if we don't serialize the models (i.e. assign post ID to job, get from DB when job runs)... I assume we want to?

### CODE REVIEW

**Comment by luceos** [comment]
Yeah would be better to load the model anew in the job. You can also load the latest version with a method, fresh? Need to look at that from a computer.

**Comment by dsevillamartin** [comment]
@luceos How it is set up right now with `SerializesModels` is that it gets it from DB, and doesn't save all the properties in the queue job (if using i.e. redis). Could also use `fresh` but that would save all the properties, so it's basically the same thing but adding more useless data to the job.

**Comment by luceos** [comment]
Right the queue reads the entry from the database again on deserialization.

**Comment by franzliedke** [comment]
Do we want to sunset this in favor of flarum/core#1931? Does that one have the problem with the last post number?

**Comment by dsevillamartin** [comment]
Yeah, I assume merging flarum/core#1931 would result in the problem that is also introduced here if I didn't add the last post number parameter. I encountered the error with follow-tags before making this PR, which is why I knew I needed it.

If the queue job runs after a new post is made, the last post number will be that one instead of the one that dispatched the notifications into the queue. So I would assume it would cause the same problem.

**Comment by franzliedke** [comment]
@datitisev Can you help @luceos with the last (small) outstanding issue in flarum/core#1931? And if that gets merged, this PR will essentially become a bugfix (so as to only fix the post number thing), right? But it should be smaller then...

**Comment by franzliedke** [comment]
Never mind, I think I understand now (probably for the first time, lol). For this particular notification, we need to move the call to `NotificationSyncer->sync()` into the queue, which is exactly what this does.

But does this still have a problem with the `async` queue or not? :confused:

**Comment by dsevillamartin** [comment]
@franzliedke This PR fixes the problem with the `async` queue of not sending notifications as the check for the latest post being read would include the one being notified above, so it would be false.

TL;DR: notifs won't be sent on `async` queues unless jobs run after user has already viewed post

---

About the small outstanding issue in the PR, not sure exactly what you want to do there / how to fix it, my apologies.

**Comment by luceos** [comment]
I think we should merge this as is. Moving the blueprint to be synced by the NotificationSyncer directly will cause the query to run before the queue processing. This solution moves that out of the process. Did you have any other feedback to block the PR @franzliedke ?

PS I tested this locally and it works fine.

**Comment by franzliedke** [comment]
I think this is sensible, too; I think it was just lying around waiting for the core PR. Thanks for diving into this, @datitisev!
